pkg/csi/service/vanilla/controller_test.go

```diff
@@ -773,3 +774,83 @@ func TestCompleteControllerFlow(t *testing.T) {
 		t.Fatalf("Volume should not exist after deletion with ID: %s", volID)
 	}
 }
+
+func TestDeleteVolumeWithSnapshots(t *testing.T) {
+	ct := getControllerTest(t)
+
+	// Create.
+	params := make(map[string]string)
+	if v := os.Getenv("VSPHERE_DATASTORE_URL"); v != "" {
+		params[common.AttributeDatastoreURL] = v
+	}
+	capabilities := []*csi.VolumeCapability{
+		{
+			AccessMode: &csi.VolumeCapability_AccessMode{
+				Mode: csi.VolumeCapability_AccessMode_SINGLE_NODE_WRITER,
+			},
+		},
+	}
+
+	reqCreate := &csi.CreateVolumeRequest{
+		Name: testVolumeName + "-" + uuid.New().String(),
+		CapacityRange: &csi.CapacityRange{
+			RequiredBytes: 1 * common.GbInBytes,
+		},
+		Parameters:         params,
+		VolumeCapabilities: capabilities,
+	}
+
+	respCreate, err := ct.controller.CreateVolume(ctx, reqCreate)
+	if err != nil {
+		t.Fatal(err)
+	}
+	volID := respCreate.Volume.VolumeId
+
+	// Create Snapshot
+	snapSpec := cnstypes.CnsSnapshotCreateSpec{
+		VolumeId: cnstypes.CnsVolumeId{
+			Id: volID,
+		},
+		Description: "Snap-1",
+	}
+	snapshotTask, err := ct.vcenter.CnsClient.CreateSnapshots(ctx, []cnstypes.CnsSnapshotCreateSpec{snapSpec})
+	if err != nil {
+		t.Fatalf("failed to create snapshot for volume: %s", volID)
+	}
+	snapTaskInfo, err := cns.GetTaskInfo(ctx, snapshotTask)
+	if err != nil {
+		t.Fatalf("failed to get snapshot taskinfo for volume: %s", volID)
+	}
+	taskResult, err := cns.GetTaskResult(ctx, snapTaskInfo)
+	if err != nil {
+		t.Fatalf("failed to get task result for snapshot operation on volume: %s", volID)
+	}
+	cnsSnapshot := taskResult.(*cnstypes.CnsSnapshotCreateResult)
+	snapshotID := cnsSnapshot.Snapshot.SnapshotId.Id
+	t.Logf("created cns snapshot :%s for volume :%s", snapshotID, volID)
+	defer func() {
+		deleteSnapSepc := cnstypes.CnsSnapshotDeleteSpec{
+			VolumeId: cnstypes.CnsVolumeId{
+				Id: volID,
+			},
+			SnapshotId: cnstypes.CnsSnapshotId{
+				Id: snapshotID,
+			},
+		}
+		_, err = ct.vcenter.CnsClient.DeleteSnapshots(ctx, []cnstypes.CnsSnapshotDeleteSpec{deleteSnapSepc})
+		if err != nil {
+			t.Errorf("failed to delete snapshot %s for volume %s", snapshotID, volID)
+		}
+	}()
+
+	//Attempt to delete the volume
+	reqDelete := &csi.DeleteVolumeRequest{
+		VolumeId: volID,
+	}
+	_, err = ct.controller.DeleteVolume(ctx, reqDelete)
+	if err != nil {
+		t.Logf("successfully received error when attempting to delete volume with snapshot, err: %+v", err)
```

Comment:
This message reads weird.  Isn't it normal to get error message if err != nil?

Reply:
Same explanation as above.

Reply:
would it be better to compare the returned error code as well?

Reply:
Done.